zerver/lib/subscription_info.py

```diff
@@ -336,23 +337,50 @@ def validate_user_access_to_subscribers(user_profile: UserProfile | None, stream
     * The stream is invite only, requesting_user is passed, and that user
       does not subscribe to the stream.
     """
+    user_recursive_group_ids: set[int] = set()
+    # Optimization for the organization administrator code path. We
+    # don't explicitly grant realm admins this permission, but admins
+    # implicitly have the can_administer_channel_group permission for
+    # all channels. user_recursive_group_ids is used to check the
+    # membership of the current user in can_administer_channel_group
+    # which we don't need to calculate in case of a realm admin.
+    if user_profile and not user_profile.is_realm_admin:
+        user_recursive_group_ids = set(
+            get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+        )
+
     validate_user_access_to_subscribers_helper(
         user_profile,
         {
             "realm_id": stream.realm_id,
             "is_web_public": stream.is_web_public,
             "invite_only": stream.invite_only,
+            "can_administer_channel_group_id": stream.can_administer_channel_group_id,
+            "can_add_subscribers_group_id": stream.can_add_subscribers_group_id,
         },
         # We use a lambda here so that we only compute whether the
         # user is subscribed if we have to
         lambda user_profile: subscribed_to_stream(user_profile, stream.id),
+        user_recursive_group_ids,
+    )
+
+
+def has_metadata_access_to_channel(
+    user_recursive_group_ids: set[int],
+    can_administer_channel_group_id: int,
+    can_add_subscribers_group_id: int,
+) -> bool:
+    return (
+        can_administer_channel_group_id in user_recursive_group_ids
+        or can_add_subscribers_group_id in user_recursive_group_ids
     )
```

Comment:
This function doesn't consider whether the user is subscribed or an organization administrator. Either it should be renamed (maybe `channel_permissions_grant_metadata_access` or include those in its logic.